swiftsimio/accelerated.py

```diff
@@ -168,3 +170,271 @@ def index_dataset(handle: Dataset, mask_array: np.array) -> np.array:
+        # that it doesn't already exist or overlap with an
+        # existing one. The only way overlap can happen is
+        # if the current lower index is less than or equal
+        # to the previous upper one. In that case simply
+        # update the previous upper to cover current chunk
+        if len(chunk_ranges) > 0:
+            if lower > chunk_ranges[-1][1]:
+                chunk_ranges.append([lower, upper])
+            elif lower <= chunk_ranges[-1][1]:
+                chunk_ranges[-1][1] = upper
+        # If chunk_ranges is empty, don't do any checks
+        else:
+            chunk_ranges.append([lower, upper])
+
+    return np.asarray(chunk_ranges)
+
+
+@jit(nopython=True, fastmath=True)
+def expand_ranges(ranges: np.ndarray) -> np.array:
+    """
+    Return an array of indices that are within the specified ranges
+
+    Parameters
+    ----------
+    ranges : np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    Returns
+    -------
+    np.array
+        1D array of indices that fall within each range specified in `ranges`
+        
+    """
+    length = np.asarray([bounds[1] - bounds[0] for bounds in ranges]).sum()
+
+    output = np.zeros(length, dtype=np.int64)
+    i = 0
+    for bounds in ranges:
+        lower = bounds[0]
+        upper = bounds[1]
+        bound_length = upper - lower
+        output[i : i + bound_length] = np.arange(lower, upper, dtype=np.int64)
+        i += bound_length
+
+    return output
+
+
+@jit(nopython=True, fastmath=True)
+def extract_ranges_from_chunks(
+    array: np.ndarray, chunks: np.ndarray, ranges: np.ndarray
+) -> np.ndarray:
+    """
+    Returns elements from array that are located within specified ranges
+    
+    `array` is a portion of the dataset being read consisting of all the chunks
+    that contain the ranges specified in `ranges`. The `chunks` array contains
+    the indices of the upper and lower bounds of these chunks. To find the 
+    elements of the dataset that lie within the specified ranges we first create
+    an array indexing which chunk each range belongs to. From this information 
+    we create an array of adjusted ranges that takes into account that the array
+    is not the whole dataset. We then return the values in `array` that are 
+    within the adjusted ranges.
+    
+    Parameters
+    ----------
+    array : np.ndarray
+        array containing data read in from snapshot
+    chunks : np.ndarray
+        two dimensional array of bounds for the chunks that contain each range from
+        `ranges`
+    ranges: np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    Returns
+    -------
+    np.ndarray
+        subset of `array` whose elements are within each range in `ranges`
+    
+    """
+    # Find out which of the chunks in the chunks array each range in ranges belongs to
+    n_ranges = len(ranges)
+    n_chunks = len(chunks)
+    chunk_array_index = np.zeros(len(ranges), dtype=np.int32)
```

Comment:
You just created the variable `n_ranges` for this exact purpose - use it instead of `len(ranges)` again.